File: frontend/src/api/mall.ts
const API_BASE = import.meta.env.VITE_API_BASE ?? '';

export type NutritionRow = {
  label: string;
  value: string;
};

export type MallProduct = {
  product_id: string;
  name: string;
  brand: string | null;
  category_code: string;
  category_name: string;
  price_cents: number;
  price_text: string;
  original_price_cents: number | null;
  original_price_text: string | null;
  spec: string | null;
  sales_text: string | null;
  image_emoji: string | null;
  image_url: string | null;
  health_tags: string[];
  recommend_reason: string | null;
};

export type MallProductDetail = MallProduct & {
  description: string | null;
  ingredients: string | null;
  shelf_life: string | null;
  nutrition_rows: NutritionRow[];
  warning_tags: string[];
};

export type MallZone = {
  zone_code: string;
  name: string;
  zone_type: string;
  icon: string | null;
  match_tag: string | null;
  sort_order: number;
};

export type MallFamilyRecommendation = {
  member_id: string;
  member_name: string;
  relation: string;
  zone_name: string;
  summary: string;
  products: MallProduct[];
};

export type MallHomeResponse = {
  family_recommendations: MallFamilyRecommendation[];
  health_zones: MallZone[];
  daily_products: MallProduct[];
  categories: MallZone[];
};

export type MallProductListResponse = {
  products: MallProduct[];
  zone: MallZone | null;
};

export type MallProductDetailResponse = {
  product: MallProductDetail;
  recommend_reason: string | null;
  nutrition_rows: NutritionRow[];
  related_products: MallProduct[];
  health_notice: string;
};

export type MallCartItem = {
  product_id: string;
  name: string;
  spec: string | null;
  image_emoji: string | null;
  image_url: string | null;
  price_cents: number;
  price_text: string;
  quantity: number;
  subtotal_cents: number;
  subtotal_text: string;
};

export type MallCartResponse = {
  items: MallCartItem[];
  total_quantity: number;
  total_cents: number;
  total_text: string;
};

async function readJson<T>(response: Response, fallback: string): Promise<T> {
  const contentType = response.headers.get('content-type') ?? '';
  if (!contentType.includes('application/json')) {
    throw new Error(fallback);
  }
  if (!response.ok) {
    const detail = await response.json().catch(() => null);
    throw new Error(detail?.detail ?? fallback);
  }
  return response.json();
}

export async function getMallHome(): Promise<MallHomeResponse> {
  const response = await fetch(`${API_BASE}/mall/home`);
  return readJson<MallHomeResponse>(response, '获取商城首页失败');
}

export async function listMallProducts(params?: {
  zone_code?: string;
  category_code?: string;
  member_id?: string;
  limit?: number;
}): Promise<MallProductListResponse> {
  const search = new URLSearchParams();
  if (params?.zone_code) search.set('zone_code', params.zone_code);
  if (params?.category_code) search.set('category_code', params.category_code);
  if (params?.member_id) search.set('member_id', params.member_id);
  if (params?.limit) search.set('limit', String(params.limit));
  const query = search.toString();
  const url = `${API_BASE}/mall/products${query ? `?${query}` : ''}`;
  const response = await fetch(url);
  return readJson<MallProductListResponse>(response, '获取商品列表失败');
}

export async function getMallProduct(productId: string): Promise<MallProductDetailResponse> {
  const response = await fetch(`${API_BASE}/mall/products/${productId}`);
  return readJson<MallProductDetailResponse>(response, '获取商品详情失败');
}

export async function getMallCart(): Promise<MallCartResponse> {
  const response = await fetch(`${API_BASE}/mall/cart`);
  return readJson<MallCartResponse>(response, '获取购物车失败');
}

export async function addMallCartItem(productId: string, quantity: number): Promise<MallCartResponse> {
  const response = await fetch(`${API_BASE}/mall/cart/items`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ product_id: productId, quantity })
  });
  return readJson<MallCartResponse>(response, '加入购物车失败');
}

export async function updateMallCartItem(productId: string, quantity: number): Promise<MallCartResponse> {
  const response = await fetch(`${API_BASE}/mall/cart/items/${productId}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ quantity })
  });
  return readJson<MallCartResponse>(response, '修改购物车失败');
}

export async function deleteMallCartItem(productId: string): Promise<void> {
  const response = await fetch(`${API_BASE}/mall/cart/items/${productId}`, {
    method: 'DELETE'
  });
  if (!response.ok) {
    const detail = await response.json().catch(() => null);
    throw new Error(detail?.detail ?? '删除购物车商品失败');
  }
}

File: backend/app/schemas/mall.py
import json

from pydantic import BaseModel, Field, computed_field


def _safe_json_list(raw: str | None) -> list:
    if not raw:
        return []
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return []
    if not isinstance(value, list):
        return []
    return value


def _format_price(cents: int) -> str:
    yuan = cents / 100
    if yuan == int(yuan):
        return f"¥{int(yuan)}"
    return f"¥{yuan:.1f}"


class NutritionRow(BaseModel):
    label: str
    value: str


class MallProductSummary(BaseModel):
    product_id: str
    name: str
    brand: str | None = None
    category_code: str
    category_name: str
    price_cents: int
    original_price_cents: int | None = None
    spec: str | None = None
    sales_text: str | None = None
    image_emoji: str | None = None
    image_url: str | None = None
    health_tags_raw: str | None = None
    recommend_reason: str | None = None

    @computed_field
    @property
    def price_text(self) -> str:
        return _format_price(self.price_cents)

    @computed_field
    @property
    def original_price_text(self) -> str | None:
        if self.original_price_cents is None:
            return None
        return _format_price(self.original_price_cents)

    @computed_field
    @property
    def health_tags(self) -> list[str]:
        return _safe_json_list(self.health_tags_raw)


class MallProductDetail(MallProductSummary):
    description: str | None = None
    ingredients: str | None = None
    shelf_life: str | None = None
    nutrition_raw: str | None = None
    warning_tags_raw: str | None = None

    @computed_field
    @property
    def nutrition_rows(self) -> list[NutritionRow]:
        rows = _safe_json_list(self.nutrition_raw)
        result = []
        for item in rows:
            if isinstance(item, dict) and "label" in item and "value" in item:
                result.append(NutritionRow(label=item["label"], value=item["value"]))
        return result

    @computed_field
    @property
    def warning_tags(self) -> list[str]:
        return _safe_json_list(self.warning_tags_raw)


class MallZone(BaseModel):
    zone_code: str
    name: str
    zone_type: str
    icon: str | None = None
    match_tag: str | None = None
    sort_order: int = 0


class MallFamilyRecommendation(BaseModel):
    member_id: str
    member_name: str
    relation: str
    zone_name: str
    summary: str
    products: list[MallProductSummary] = Field(default_factory=list)


class MallHomeResponse(BaseModel):
    family_recommendations: list[MallFamilyRecommendation] = Field(default_factory=list)
    health_zones: list[MallZone] = Field(default_factory=list)
    daily_products: list[MallProductSummary] = Field(default_factory=list)
    categories: list[MallZone] = Field(default_factory=list)


class MallProductListResponse(BaseModel):
    products: list[MallProductSummary] = Field(default_factory=list)
    zone: MallZone | None = None


class MallProductDetailResponse(BaseModel):
    product: MallProductDetail
    recommend_reason: str | None = None
    nutrition_rows: list[NutritionRow] = Field(default_factory=list)
    related_products: list[MallProductSummary] = Field(default_factory=list)
    health_notice: str = "本推荐不构成医疗建议"


class MallCartItem(BaseModel):
    product_id: str
    name: str
    spec: str | None = None
    image_emoji: str | None = None
    image_url: str | None = None
    price_cents: int
    quantity: int
    subtotal_cents: int = 0

    @computed_field
    @property
    def price_text(self) -> str:
        return _format_price(self.price_cents)

    @computed_field
    @property
    def subtotal_text(self) -> str:
        return _format_price(self.subtotal_cents)


class MallCartResponse(BaseModel):
    items: list[MallCartItem] = Field(default_factory=list)
    total_quantity: int = 0
    total_cents: int = 0

    @computed_field
    @property
    def total_text(self) -> str:
        return _format_price(self.total_cents)


class MallCartItemCreateRequest(BaseModel):
    product_id: str
    quantity: int = 1


class MallCartItemUpdateRequest(BaseModel):
    quantity: int
